rllib/core/rl_trainer/tf/tf_rl_trainer.py

```diff
@@ -0,0 +1,150 @@
+        with tf.GradientTape() as tape:
+            fwd_out = self._module.forward_train(batch)
+            loss = self._rl_optimizer.compute_loss(batch, fwd_out)
+            if isinstance(loss, tf.Tensor):
+                loss = {"total_loss": loss}
+        gradients = self.compute_gradients(loss, tape)
+        self.apply_gradients(gradients)
+        return {"loss": loss, "fwd_out": fwd_out, "post_processed_gradients": gradients}
+
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        # TODO(sven): This is a hack to get around the fact that
+        # SampleBatch.count becomes 0 after decorating the function with
+        # tf.function. This messes with input spec checking. Other fields of
+        # the sample batch are possibly modified by tf.function which may lead
+        # to unwanted consequences. We'll need to further investigate this.
+        batch = NestedDict(batch.policy_batches)
+        for key, value in batch.items():
+            batch[key] = tf.convert_to_tensor(value, dtype=tf.float32)
+        if self.distributed:
+            infos = self.strategy.run(self._do_update_fn, args=(batch,))
+        else:
+            infos = self._do_update_fn(batch)
+        loss = infos["loss"]
+        fwd_out = infos["fwd_out"]
+        post_processed_gradients = infos["post_processed_gradients"]
+        results = self.compile_results(batch, fwd_out, loss, post_processed_gradients)
+        return results
+
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]], tape: tf.GradientTape
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            loss: variable(s) used for optimizing self._module.
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        trainable_variables = {
+            module_id: self._module[module_id].trainable_variables()
+            for module_id in self._module.keys()
+        }
+        grads = tape.gradient(loss["total_loss"], trainable_variables)
+        return grads
+
+    def apply_gradients(self, gradients: Mapping[str, Any]) -> None:
+        """Perform an update on self._module"""
+        trainable_variables = {
+            module_id: self._module[module_id].trainable_variables()
+            for module_id in self._module.keys()
+        }
+        for module_id, rl_optimizer in self._rl_optimizer.get_optimizers().items():
+            for key, optimizer in rl_optimizer.get_optimizers().items():
+                optimizer.apply_gradients(
+                    zip(gradients[module_id][key], trainable_variables[module_id][key])
+                )
+
+    def make_distributed(self):
```

Comment:
This should be overriding some parent method

Reply:
done